src/delphi/train/tokenized_chunks_dataset.py

```diff
@@ -0,0 +1,57 @@
+import torch
+from torch.utils.data import Dataset
+
+from delphi.train.shuffle import shuffle_list
+
+
+class TokenizedChunksDataset(Dataset):
+    def __init__(self, tokenized_docs, max_seq_len, device):
+        self.device = device
+        self.tokenized_docs = tokenized_docs
+        self.doc_len = len(tokenized_docs[0]["tokens"])
+        self.max_len = max_seq_len
+        self._total_tokens = self.doc_len * len(self.tokenized_docs)
+        self.batched_tokens = (
+            torch.Tensor()
+        )  # will be initialized in initialize_samples
+
+    def initialize_samples(self):
+        # self.tokenized_docs is an (X, 1) tensor of dicts. Each entry is just {"tokens": [int]}
+        # where [int] is doc_len long
+        # we want to turn this into a (num_batches, max_len + 1) tensor of ints
+        # the +1 is for the last Y token prediction, and implies an overlap of 1 token between batches
+        # this is because each batch will be broken into X [:-1] and Y [1:]
+        tensor_tokens = torch.stack(
+            [torch.tensor(doc["tokens"]) for doc in self.tokenized_docs]
+        ).to(self.device)
+        self.batched_tokens = tensor_tokens.flatten().unfold(
+            0, self.max_len + 1, self.max_len
+        )
+        self.indices = self._default_indices()
+
+    def _default_indices(self):
+        return list(range(len(self.batched_tokens)))
+
+    def shuffle(self, epoch: int):
```

Comment:
I don't know if this is correct. Shuffle alters the list in place, why are we changing the indices instead of the list?

Reply:
Idempotency. I want `shuffle(x)` to result in the same state regardless of when it's called - so not dependent on the state of the shuffling before it's called. By making it a deterministic shuffle of the default indices we can guarantee that it doesn't matter when you call `shuffle(x)`, you'll still get the same result.

Technically we don't need this for reproducibility if we commit to always calling shuffle with the same arguments in the same sequence, but I prefer to have it work this way to minimize the amount of state that needs to be kept track of when debugging - I don't want to have to repeat every shuffle in a given sequence to reproduce a problem.

Reply:
I like it that way

Reply:
I just didnt see where you then apply the indices to the list, but if you do it somewhere I trust you